// quip-apps-api/dist/blob.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Blob {
    constructor() {
        this.hasLoadedDataValue = false;
        this.imageHeightValue = 0;
        this.imageWidthValue = 0;
    }
    getId() {
        return this.idValue;
    }
    id() {
        return this.getId();
    }
    data() {
        return this.getData();
    }
    getData() {
        return this.dataValue;
    }
    downloadAsFile() { }
    filename() {
        return this.getFilename();
    }
    getFilename() {
        return this.filenameValue;
    }
    hasLoadedData() {
        return this.hasLoadedDataValue;
    }
    imageHeight() {
        return this.imageHeightValue;
    }
    imageWidth() {
        return this.imageWidthValue;
    }
    onDataLoaded(loadedCallback, failedCallback) { }
    openInLightbox() { }
    url() {
        // TODO: not sure this type coercion is always safe
        return URL.createObjectURL(new window.Blob([this.dataValue]));
    }
}
exports.default = Blob;
