// quip-apps-api/dist/image-record.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const record_1 = __importDefault(require("./record"));
const canvas_record_1 = __importDefault(require("./canvas-record"));
class ImageRecord extends record_1.default {
    constructor() {
        super(...arguments);
        this.canvasRecordValue = new canvas_record_1.default();
        this.originalImageHeightValue = 0;
        this.originalImageWidthValue = 0;
        this.hasImageValue = false;
    }
    chooseFile() { }
    commitCrop(fractionLeft, fractionTop, fractionWidth, fractionHeight, cropInitiatedKey) { }
    downloadOriginal() { }
    getCanvasRecord() {
        return this.canvasRecordValue;
    }
    getOriginalImageHeight() {
        return this.originalImageHeightValue;
    }
    getOriginalImageWidth() {
        return this.originalImageWidthValue;
    }
    handleShowFilePickerAction(blobsWithThumbnails) { }
    handleShowFilePickerActionStarted() { }
    hasImage() {
        return this.hasImageValue;
    }
    openInLightbox() { }
    uploadFile(blob) { }
}
exports.default = ImageRecord;
