// quip-apps-api/dist/quip.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
const auth_1 = __importDefault(require("./auth"));
const base_oauth_1 = __importDefault(require("./base-oauth"));
const blob_1 = __importDefault(require("./blob"));
const canvas_record_1 = __importDefault(require("./canvas-record"));
const client_1 = __importStar(require("./client")), clientExports = client_1;
const client_error_1 = __importDefault(require("./client-error"));
const image_record_1 = __importDefault(require("./image-record"));
const oauth_1_1 = __importDefault(require("./oauth-1"));
const oauth_2_1 = __importDefault(require("./oauth-2"));
const preferences_1 = __importDefault(require("./preferences"));
const record_1 = __importDefault(require("./record"));
const record_index_1 = __importDefault(require("./record-index"));
const record_list_1 = __importDefault(require("./record-list"));
const register_migration_1 = __importDefault(require("./register-migration"));
const rich_text_record_1 = __importDefault(require("./rich-text-record"));
const root_record_1 = __importDefault(require("./root-record"));
const ui = __importStar(require("./ui"));
const url_auth_1 = __importDefault(require("./url-auth"));
const user_1 = __importDefault(require("./user"));
const client = new client_1.default("", "", "");
const api = Object.assign(client, clientExports, {
    Auth: auth_1.default,
    BaseOAuth: base_oauth_1.default,
    Blob: blob_1.default,
    CanvasRecord: canvas_record_1.default,
    Record: record_1.default,
    Entity: record_1.default,
    ChildEntity: record_1.default,
    ClientError: client_error_1.default,
    ImageRecord: image_record_1.default,
    ImageEntity: image_record_1.default,
    OAuth1: oauth_1_1.default,
    OAuth2: oauth_2_1.default,
    Preferences: preferences_1.default,
    RecordIndex: record_index_1.default,
    RootRecord: root_record_1.default,
    RootEntity: root_record_1.default,
    RecordList: record_list_1.default,
    RichTextRecord: rich_text_record_1.default,
    RichTextEntity: rich_text_record_1.default,
    UrlAuth: url_auth_1.default,
    User: user_1.default,
    ui,
    registerMigration: register_migration_1.default,
});
const quip = {
    apps: api,
    elements: api,
};
module.exports = quip;
