// quip-apps-api/dist/rich-text-record.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const record_1 = __importDefault(require("./record"));
class RichTextRecord extends record_1.default {
    constructor() {
        super(...arguments);
        this.emptyValue = false;
        this.placeholderTextValue = "";
        this.textContentValue = "";
    }
    clearContent() { }
    delete() {
        this.setDeleted();
    }
    empty() {
        return this.emptyValue;
    }
    focus(fromPrevious) { }
    getPlaceholderText() {
        return this.placeholderTextValue;
    }
    getTextContent() {
        return this.textContentValue;
    }
    insertText(text) { }
    listenToContent(listener) { }
    unlistenToContent(listener) { }
    listenToSave(listener) { }
    unlistenToSave(listener) { }
    replaceContent(text) { }
    selectAll() { }
    setDeleted() {
        this.isDeletedValue = true;
    }
}
exports.default = RichTextRecord;
RichTextRecord.InlineStyle = {
    BOLD: 1,
    ITALIC: 2,
    UNDERLINE: 3,
    STRIKETHROUGH: 4,
    CODE: 5,
};
RichTextRecord.Style = {
    TEXT_PLAIN: 1,
    TEXT_H1: 2,
    TEXT_H2: 3,
    TEXT_H3: 4,
    TEXT_CODE: 5,
    TEXT_BLOCKQUOTE: 6,
    TEXT_PULL_QUOTE: 7,
    LIST_BULLET: 8,
    LIST_NUMBERED: 9,
    LIST_CHECKLIST: 10,
    HORIZONTAL_RULE: 11,
    IMAGE: 12,
};
